lib/web_catalog/api_extractor.es6.js

```diff
@@ -0,0 +1,279 @@
+    blacklistObjectAttr.push(cleanUp(objectProp[i]));
+  }
+}
+
+// Search object id's prototype chain to find the
+// class it belongs to. The class is the prototype
+// with a contructor property.
+function getClassName(id, og) {
+  var ctorId = og.lookup('+constructor+', id);
+  if (ctorId) {
+    return og.getFunctionName(ctorId);
+  }
+  return og.getFunctionName(og.getPrototype(id));
+}
+
+// Remove blacklist properties and add Function and Object's APIs.
+function postProcess(apiCatalogs, og) {
+  for (let i = 0; i < blacklistProperties.length; i += 1) {
+    var prop = blacklistProperties[i];
+    if (apiCatalogs[prop]) {
+      delete apiCatalogs[prop];
+    }
+  }
+  // Add function and obejct blacklisted property
+  // to Function and Object interface.
+  if (!apiCatalogs[og.getFunctionName(FunctionId)]) {
+    apiCatalogs[og.getFunctionName(FunctionId)] = [];
+  }
+  apiCatalogs[og.getFunctionName(FunctionId)] =
+    apiCatalogs[og.getFunctionName(FunctionId)].concat(blacklistFunctionAttr);
+  if (!apiCatalogs[og.getFunctionName(ObjectId)]) {
+    apiCatalogs[og.getFunctionName(ObjectId)] = [];
+  }
+  apiCatalogs[og.getFunctionName(ObjectId)] =
+    apiCatalogs[og.getFunctionName(ObjectId)].concat(blacklistObjectAttr);
+}
+
+/** find API for interface of something with given id.
+  @param {ObjectGraph} og - The obejct graph.
+  @param {JSON} map - the json to store web api for interface
+  @param {int} id - the id currently visiting
+  @param {String} name - the property name from its parent node. (not necessarily interface name)
+  @param {Array} visited - a list of visted properties
+  @param {JSON} opt - know if this object is firstlevel or a __proto__ property.
+*/
+function extractAPI(og, map, id, name, visited, opt) {
+  opt = opt || {};
+  var firstLevel = opt.firstLevel || false;
+  var proto = opt.proto || false;
+
+  // Return if this object is already visited, except for first level objects.
+  // Still visit it if this is a firstLevel object, since
+  // two different firstlevel object can reference to one object.
+  // We don't want to miss any of them.
+  if (visited.indexOf(id) !== -1 && !firstLevel) return;
+  // An object/function's Id must be greater than root's Id.
+  // And we don't want to revisit window object.
+  if (id <= rootId) return;
+  visited.push(id);
+  if (og.isFunction(id) && (proto || firstLevel)) {
+    // This object is a function and it is either a __proto__ or on first level.
+    // Function which is not a __protot__ or first level is not considered as an interface
+    // eg: chrome.webstore, chrome.runtime.
+    if (id === functionId) return; // Skip if function is Function.prototype;
+    let prototypeId = og.lookup('prototype', id);
+    let ownProperties = getObjectProperties(og, id);
+    let prototypeProps = [];
+    // If function exists as an first level interface, use its property name
+    // under window object as interface name, otherwise, use function name.
```

Comment:
Two first level object may reference to the same object. For example, window.webkitMediaStream and window.MediaStream are same function. If only use function name, webkitMediaStream interface will not be captured.